src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -13,24 +12,28 @@ private static void HandleStartObject(JsonSerializerOptions options, ref ReadSta
         {
             Debug.Assert(!state.Current.IsProcessingDictionary());
 
+            // Note: unless we are a root object, we are going to push a property onto the ReadStack
+            // in the if/else if check below.
+
             if (state.Current.IsProcessingEnumerable())
             {
-                // A nested object within an enumerable.
+                // A nested object within an enumerable (non-dictionary).
 
                 if (!state.Current.CollectionPropertyInitialized)
                 {
                     // We have bad JSON: enumerable element appeared without preceding StartArray token.
-                    Debug.Assert(state.Current.JsonPropertyInfo != null);
                     ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(state.Current.JsonPropertyInfo.DeclaredPropertyType);
                 }
 
                 Type objType = state.Current.GetElementType();
                 state.Push();
                 state.Current.Initialize(objType, options);
             }
-            else if (state.Current.IsProcessingObject(ClassType.Object) && state.Current.JsonPropertyInfo != null)
```

Comment:
We don't need the first clause of this check because we wouldn't be in this method if false.